planner/core/find_best_task.go

```diff
@@ -416,6 +433,186 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty) (t task, err
+	}
+	isCovered := isCoveringIndex(ds.schema.Columns, path.index.Columns, ds.tableInfo.PKIsHandle)
+	indexConds := path.indexFilters
+	if indexConds != nil {
+		var selectivity float64
+		partialCost += rowCount * copCPUFactor
+		if path.countAfterAccess > 0 {
+			selectivity = path.countAfterIndex / path.countAfterAccess
+		}
+		count := is.stats.RowCount * selectivity
+		stats := &property.StatsInfo{RowCount: count}
+		stats.StatsVersion = ds.statisticTable.Version
+		if ds.statisticTable.Pseudo {
+			stats.StatsVersion = statistics.PseudoVersion
+		}
+		indexSel := PhysicalSelection{Conditions: indexConds}.Init(is.ctx, stats)
+		indexSel.SetChildren(is)
+		partialCost += count * rowSize * netWorkFactor
+		return indexSel, partialCost, count, isCovered
+	}
+	return is, partialCost, rowCount, isCovered
+}
+
+func (ds *DataSource) convertToPartialTableScan(prop *property.PhysicalProperty, path *accessPath) (PhysicalPlan, float64, float64, bool) {
+	var partialCost float64
+	ts := PhysicalTableScan{
+		Table:           ds.tableInfo,
+		Columns:         ds.Columns,
+		TableAsName:     ds.TableAsName,
+		DBName:          ds.DBName,
+		isPartition:     ds.isPartition,
+		physicalTableID: ds.physicalTableID,
+	}.Init(ds.ctx)
+	ts.SetSchema(ds.schema)
+	if ts.Table.PKIsHandle {
+		if pkColInfo := ts.Table.GetPkColInfo(); pkColInfo != nil {
+			if ds.statisticTable.Columns[pkColInfo.ID] != nil {
+				ts.Hist = &ds.statisticTable.Columns[pkColInfo.ID].Histogram
+			}
+		}
+	}
+	ts.Ranges = path.ranges
+	ts.AccessCondition, ts.filterCondition = path.accessConds, path.tableFilters
+	rowCount := path.countAfterAccess
+	// Adjust number of rows we actually need to scan if prop.ExpectedCnt is smaller than the count we calculated.
+	if prop.ExpectedCnt < ds.stats.RowCount {
+		count, ok, corr := ds.crossEstimateRowCount(path, prop.ExpectedCnt, false)
+		if ok {
+			rowCount = count
+		} else if corr < 1 {
+			correlationFactor := math.Pow(1-corr, float64(ds.ctx.GetSessionVars().CorrelationExpFactor))
+			selectivity := ds.stats.RowCount / rowCount
+			rowCount = math.Min(prop.ExpectedCnt/selectivity/correlationFactor, rowCount)
+		}
+	}
+	ts.stats = ds.tableStats.ScaleByExpectCnt(rowCount)
+	rowSize := ds.TblColHists.GetAvgRowSize(ds.TblCols, false)
+	ts.stats.StatsVersion = ds.statisticTable.Version
+	if ds.statisticTable.Pseudo {
+		ts.stats.StatsVersion = statistics.PseudoVersion
+	}
+	partialCost += rowCount * rowSize * scanFactor
+	partialCost += rowCount * rowSize * netWorkFactor
+	if len(ts.filterCondition) > 0 {
+		sel := PhysicalSelection{Conditions: ts.filterCondition}.Init(ts.ctx, ts.stats)
+		sel.SetChildren(ts)
+		partialCost += rowCount * copCPUFactor
+		return sel, partialCost, rowCount, true
+	}
+	return ts, partialCost, rowCount, true
+}
+
+func (ds *DataSource) buildIndexMergeTableScan(prop *property.PhysicalProperty, tableFilters []expression.Expression, totalRowCount float64) (PhysicalPlan, float64) {
```

Comment:
This function shares a lot of code with `convertToPartialTableScan`, can we extract a common utility function for reuse?

Reply:
And also with `convertToTableScan`, `convertToPartialIndexScan`, it is better to use small functions that can compose rather than long and repeated codes.

Reply:
PTAL @hailanwhu